hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/ConcurrentInboundEdgeStream.java

```diff
@@ -165,6 +177,51 @@ public ProgressState drainTo(Predicate<Object> dest) {
         return tracker.toProgressState();
     }
 
+    private ProgressState drainToWithComparator(Predicate<Object> dest) {
+        int batchSize = -1;
+        while (true) {
+            int minIndex = 0;
+            Object minItem = null;
+            for (int queueIndex = 0; queueIndex < conveyor.queueCount(); queueIndex++) {
+                final QueuedPipe<Object> q = conveyor.queue(queueIndex);
+                if (q == null) {
+                    continue;
+                }
+                Object headObject = q.peek();
+                Object headItem;
+                if (headObject instanceof ObjectWithPartitionId) {
+                    headItem = ((ObjectWithPartitionId) headObject).getItem();
+                } else {
+                    headItem = headObject;
+                }
+                if (headItem == null) {
+                    return NO_PROGRESS;
```

Comment:
You may already have made progress by consuming some items in previous `while` loop iterations.

Reply:
Done